File: cpp/source/rocketmq/Producer.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "rocketmq/Producer.h"

#include <memory>
#include <system_error>
#include <utility>

#include <spdlog/spdlog.h>

#include "ProducerImpl.h"
#include "StaticNameServerResolver.h"
#include "rocketmq/ErrorCode.h"
#include "rocketmq/SendReceipt.h"
#include "rocketmq/Transaction.h"
#include "rocketmq/RecallReceipt.h"

ROCKETMQ_NAMESPACE_BEGIN

void Producer::start() {
  impl_->start();
}

SendReceipt Producer::send(MessageConstPtr message, std::error_code& ec) noexcept {
  try {
    if (!message) {
      ec = ErrorCode::BadRequest;
      return {};
    }

    return impl_->send(std::move(message), ec);
  } catch (const std::exception& e) {
    ec = std::make_error_code(std::errc::io_error);
    SPDLOG_ERROR("Exception in send: {}", e.what());
    return {};
  }
}

void Producer::send(MessageConstPtr message, SendCallback callback) noexcept {
  try {
    if (!message) {
      std::error_code ec = ErrorCode::BadRequest;
      SendReceipt send_receipt = {};
      callback(ec, std::move(send_receipt));
      return;
    }

    if (!message->group().empty()) {
      SendReceipt     empty;
      std::error_code ec = ErrorCode::BadRequestAsyncPubFifoMessage;
      callback(ec, std::move(empty));
      return;
    }

    impl_->send(std::move(message), callback);
  } catch (const std::exception& e) {
    SPDLOG_ERROR("Exception in async send: {}", e.what());
    std::error_code ec = std::make_error_code(std::errc::io_error);
    SendReceipt empty;
    callback(ec, std::move(empty));
  }
}

std::unique_ptr<Transaction> Producer::beginTransaction() {
  return impl_->beginTransaction();
}

SendReceipt Producer::send(MessageConstPtr message, std::error_code& ec, Transaction& transaction) noexcept {
  try {
    return impl_->send(std::move(message), ec, transaction);
  } catch (const std::exception& e) {
    ec = std::make_error_code(std::errc::io_error);
    SPDLOG_ERROR("Exception in transactional send: {}", e.what());
    return {};
  }
}

RecallReceipt Producer::recall(std::string& topic, std::string& recall_handle, std::error_code& ec) noexcept {
  try {
    return impl_->recall(topic, recall_handle, ec);
  } catch (const std::exception& e) {
    ec = std::make_error_code(std::errc::io_error);
    SPDLOG_ERROR("Exception in recall: {}", e.what());
    return {};
  }
}

ProducerBuilder Producer::newBuilder() {
  return {};
}

ProducerBuilder::ProducerBuilder() : impl_(std::make_shared<ProducerImpl>()){}

ProducerBuilder& ProducerBuilder::withConfiguration(Configuration configuration) {
  auto name_server_resolver = std::make_shared<StaticNameServerResolver>(configuration.endpoints());
  impl_->withNameServerResolver(std::move(name_server_resolver));
  impl_->withResourceNamespace(configuration.resourceNamespace());
  impl_->withCredentialsProvider(configuration.credentialsProvider());
  impl_->withRequestTimeout(configuration.requestTimeout());
  impl_->withCallbackThreads(configuration.callbackThreads());
  impl_->withSsl(configuration.withSsl());
  return *this;
}

ProducerBuilder& ProducerBuilder::withTopics(std::vector<std::string> topics) {
  impl_->withTopics(std::move(topics));
  return *this;
}

ProducerBuilder& ProducerBuilder::withTransactionChecker(TransactionChecker checker) {
  impl_->transaction_checker_ = std::move(checker);
  return *this;
}

Producer ProducerBuilder::build() {
  Producer producer(impl_);
  producer.start();
  return producer;
}

ROCKETMQ_NAMESPACE_END

File: php/Producer/Producer.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq\Producer;

use Apache\Rocketmq\Exception\ClientException;
use Apache\Rocketmq\Message\Message;

/**
 * Producer is a thread-safe rocketmq client which is used to publish messages.
 *
 * <p>On account of network timeout or other reasons, the producer only promised the at-least-once semantics.
 * For the producer, at-least-once semantics means potentially attempts are made at sending it, messages may be
 * duplicated but not lost. Especially, potentially attempts are not made using {@link #send(Message, Transaction)}.
 */
interface Producer {
    /**
     * Sends a message synchronously.
     *
     * <p>This method does not return until it gets the definitive result.
     *
     * @param Message $message the message to send.
     * @param Transaction|null $transaction the transaction to bind, null for normal message.
     * @return SendReceipt the returned receipt.
     * @throws ClientException If an error occurs
     */
    public function send(Message $message, ?Transaction $transaction = null): SendReceipt;

    /**
     * Sends a message asynchronously.
     *
     * <p>This method returns immediately, the result is included in the future;
     *
     * @param Message $message the message to send.
     * @return mixed a future that indicates the send receipt.
     */
    public function sendAsync(Message $message);

    /**
     * Begins a transaction.
     *
     * <p>For example:
     *
     * <pre>{@code
     * $transaction = $producer->beginTransaction();
     * $receipt1 = $producer->send($message1, $transaction);
     * $receipt2 = $producer->send($message2, $transaction);
     * $transaction->commit();
     * }</pre>
     *
     * @return Transaction a transaction entity to execute commit/rollback operation.
     * @throws ClientException If an error occurs
     */
    public function beginTransaction(): Transaction;

    /**
     * Recall message synchronously, only delay message is supported for now.
     *
     * <pre>{@code
     * $receipt = $producer->send($message);
     * $recallHandle = $receipt->getRecallHandle();
     * }</pre>
     *
     * @param string $topic the topic of the operation
     * @param string $recallHandle the handle to identify a message to recall
     * @return RecallReceipt the returned receipt, or throw exception if response status is not OK.
     * @throws ClientException If an error occurs
     */
    public function recallMessage(string $topic, string $recallHandle): RecallReceipt;

    /**
     * Recall message asynchronously.
     *
     * <p>This method returns immediately, the result is included in the future;
     *
     * @param string $topic the topic of the operation
     * @param string $recallHandle the handle to identify a message to recall
     * @return mixed a future that indicates the receipt
     */
    public function recallMessageAsync(string $topic, string $recallHandle);

    /**
     * Closes the producer and releases all related resources.
     *
     * <p>Once producer is closed, <strong>it could not be started once again.</strong> we maintained an FSM
     * (finite-state machine) to record the different states for each producer.
     *
     * @return void
     */
    public function close();
}
